### PR TITLE
Fix Issue 21259 -  struct initialization with deprecated fields should issue deprecation warnings

### PR DESCRIPTION
**Pull request by RazvanN7**

### CODE REVIEW

File: src/dmd/initsem.d [hunk 1]
```diff
@@ -144,7 +144,13 @@ extern(C++) Initializer initializerSemantic(Initializer init, Scope* sc, Type t,
                             error(initLoc, "`%s` is not a member of `%s`", id.toChars(), sd.toChars());
                         return new ErrorInitializer();
                     }
-                    s = s.toAlias();
+                    auto s_alias = s.toAlias();
+                    if (s.isDeprecated())
+                        deprecation(i.loc, "field `%s` of struct `%s` is deprecated", s.toChars(), sd.toChars());
+                    else if (s_alias.isDeprecated())
+                        deprecation(i.loc, "field `%s` of struct `%s` is deprecated", s_alias.toChars(), sd.toChars());
```
Comment (Geod24) [inline src/dmd/initsem.d:151]: `s/field/alias/` ?

Also does this take into account deprecated scope ?

**Review by thewilsonator — APPROVED** [review]

File: src/dmd/initsem.d (hunk 1 above)
Comment (RazvanN7) [reply src/dmd/initsem.d:151]: > s/field/alias/ ?

No, think about library objects, you don't care if whether the symbol is an alias or not, you just know it is deprecated. However, I just noticed that if you have an alias to a deprecated symbol then you already get a deprecation at the alias declaration site, rendering this message redundant. I will delete the second check.

> Also does this take into account deprecated scope ? Doesn't seem to, it should probably call checkDeprecated.

Yes, it does.

**Comment by RazvanN7** [comment]
I don't think the Azure failure is related.

File: src/dmd/initsem.d (hunk 1 above)
Comment (Geod24) [reply src/dmd/initsem.d]: > Yes, it does.

Can you add a test then ?

Comment (RazvanN7) [reply src/dmd/initsem.d]: Done

**Review by wilzbach — APPROVED** [review]
Awesome! Thanks!

**Comment by Geod24** [comment]
> Also does this take into account deprecated scope ? Doesn't seem to, it should probably call checkDeprecated.


And as I suspected, it does not. The following should not trigger a deprecation:
```D
deprecated void foo ()
{
    Foo a = { width: 100 };
}
```
But with your changes, it does.

**Comment by RazvanN7** [comment]
I see what you mean now. I thought you referred to the declaration being in a deprecated scope.

**Comment by wilzbach** [comment]
-> https://github.com/dlang/dmd/pull/11755

**Pull request merged**